Knowledge distillation tutorial (#1698)
pytorch/torchtune: docs/source/tutorials/llama_kd_tutorial.rst

.. _llama_kd_label:

> ============================
> Distilling Llama3 8B into 1B
> ============================

Reviewer: ```suggestion
==================================================
Distilling Llama3.1 8B into Llama3.2 1B using Knowledge Distillation
==================================================
```

This guide will teach you about knowledge distillation (KD) and show you how you can use torchtune to distill a Llama3.1 8B model into Llama3.2 1B.
If you already know what knowledge distillation is and want to get straight to running your own distillation in torchtune,

> you can jump to knowledge distillation recipe in torchtune, `knowledge_distillation_single_device.py <https://github.com/pytorch/torchtune/blob/main/recipes/knowledge_distillation_single_device.py>`_.

Reviewer: nit
```suggestion
you can jump to the knowledge distillation recipe in torchtune, `knowledge_distillation_single_device.py <https://github.com/pytorch/torchtune/blob/main/recipes/knowledge_distillation_single_device.py>`_.
```

Reviewer: Actually, it might be better to just reference the latter parts of this tutorial for people who want to jump ahead rather than pointing to the recipe file,

Author: Changed to link the tutorial section instead of recipe.

      * Make sure to :ref:`install torchtune<install_label>`
      * Make sure you have downloaded the :ref:`Llama3 model weights<download_llama_label>`
      * Be familiar with :ref:`LoRA<lora_finetune_label>`

What is Knowledge Distillation?
-------------------------------

`Knowledge Distillation <https://arxiv.org/pdf/1503.02531>`_ is is a widely used compression technique
that transfers knowledge from a larger (teacher) model to a smaller (student) model. Larger models have
more parameters and capacity for knowledge, however, this larger capacity is also more computationally
expensive to deploy. Knowledge distillation can be used to compress the knowledge of a larger model into
a smaller model. The idea is that performance of smaller models can be improved by learning from larger
model's outputs.

How does Knowledge Distillation work?
-------------------------------------

Knowledge is transferred from the teacher to student model by training it on a transfer set and where the
student is trained to imitate the token-level probability distributions of the teacher. The diagram below
is a simplified representation of how KD works.

.. image:: /_static/img/kd-simplified.png

> The total loss can be configured in many ways. The default KD config in torchtune combines CE loss with

Reviewer: ```suggestion
The total loss can be configured in many ways. The default KD config in torchtune combines the cross-entropy (CE) loss with
```

> forward `Kullback-Leibler (KL) divergence <https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence>`_ loss,

Reviewer: ```suggestion
the forward `Kullback-Leibler (KL) divergence <https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence>`_ loss,
```

Reviewer: nit: could you link to a more concrete reference of this specific loss? e.g. https://arxiv.org/html/2402.03898v2#S2

> which is used in standard KD approaches.Forward KL divergence aims to minimize the difference by forcing the student

Reviewer: ```suggestion
which is used in standard KD approaches. iForward KL divergence aims to minimize the difference by forcing the student
```

Reviewer: which difference are you referring to? : )

the forward KL divergence loss.

.. code-block:: python

  import torch
  import torch.nn.functional as F

  class ForwardKLLoss(torch.nn.Module):
    def __init__(self, ignore_index: int = -100)
      super().__init__()
      self.ignore_index = ignore_index

    def forward(self, student_logits, teacher_logits, labels) -> torch.Tensor:
      # Implementation from https://github.com/jongwooko/distillm
      # Computes the softmax of the teacher logits
      teacher_prob = F.softmax(teacher_logits, dim=-1, dtype=torch.float32)
      # Computes the student log softmax probabilities
      student_logprob = F.log_softmax(student_logits, dim=-1, dtype=torch.float32)
      # Computes the forward KL divergence
      prod_probs = teacher_prob * student_logprob
      # Compute the sum
      x = torch.sum(prod_probs, dim=-1).view(-1)
      # We don't want to include the ignore labels in the average
      mask = (labels != self.ignore_index).int()
      # Loss is averaged over non-ignored targets
      return -torch.sum(x * mask.view(-1), dim=0) / torch.sum(mask.view(-1), dim=0)

There are some details omitted to simplify the computation, but if you'd like to know more,

> you can see the implementation in `ForwardKLLoss <https://github.com/pytorch/torchtune/blob/4234b78b914af23384ce0348f564e2119d107a96/torchtune/modules/loss/kd_losses.py>`_.

Reviewer: should be able to do
```suggestion
you can see the implementation in :class:`torchtune.modules.loss.ForwardKLLoss`.
```

> you can see the implementation in `ForwardKLLoss <https://github.com/pytorch/torchtune/blob/4234b78b914af23384ce0348f564e2119d107a96/torchtune/modules/loss/kd_losses.py>`_.

Reviewer: Is it perhaps worth mentioning that we actually use ForwardKLWithChunkedOutputLoss by default?

Author: added a note that we used the chunked loss to save memory.

The current implementation only supports student and teacher models that have the same output
logit shape and same tokenizer.

KD recipe in torchtune
----------------------

With torchtune, we can easily apply knowledge distillation to Llama3, as well as other LLM model families.

> Let's take a look at how you could distill a model using torchtune's `KD recipe <https://github.com/pytorch/torchtune/blob/4234b78b914af23384ce0348f564e2119d107a96/recipes/knowledge_distillation_single_device.py>`_.

Reviewer: Not sure if you need to link to the recipe again

> First, make sure that you have downloaded the Llama3 weights. For this example, we'll use the Llama3.1-8B as teacher and Llama3.2-1B as student.

Reviewer: ```suggestion
First, make sure that you have downloaded the Llama3.1 weights. For this example, we'll use the Llama3.1-8B as teacher and Llama3.2-1B as student.
```

.. code-block:: bash

    tune download meta-llama/Meta-Llama-3.1-8B-Instruct --output-dir /tmp/Meta-Llama-3.1-8B-Instruct --ignore-patterns "original/consolidated.00.pth" --hf_token <HF_TOKEN>

    tune download meta-llama/Llama-3.2-1B-Instruct --output-dir /tmp/Llama-3.2-1B-Instruct --ignore-patterns "original/consolidated.00.pth" --hf_token <HF_TOKEN>

> Then, we will fine-tune the teacher model with using LoRA. Based on our experiments and previous work,

Reviewer: nit nit (ignore if you like): ```suggestion
Then, we will fine-tune the teacher model with using LoRA. Based on our experiments and previous work,
```

and 1B models. To change the student model, you can first fine-tune the 1B model then modify the
student model checkpointer in the config:

.. code-block:: yaml

 checkpointer:
    _component_: torchtune.training.FullModelHFCheckpointer
    checkpoint_dir: /tmp/Llama-3.2-1B-Instruct/
    checkpoint_files: [
      hf_model_0001_0.pt
    ]

Using the fine-tuned student model boosts accuracy even further for truthfulqa, but the accuracy
drops for hellaswag and commonsense. Using a fine-tuned teacher model and baseline student
model achieved the best results on hellaswag and commonsense dataset. Based on these findings,
the best configuration will change depending on which evaluation dataset and metric you are optimizing for.

.. image:: /_static/img/kd-finetune-student.png

Based on the loss graphs, using a fine-tuned teacher model results in a lower loss irrespective of
whether the student model is fine-tuned or not. It's also interesting to note that the class loss
starts to increase when using a fine-tuned student model.

Hyperparameter tuning: learning rate
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> By default, the config has the learning rate as 3e-4, same as LoRA configs. For these experiments,

Reviewer: ```suggestion
By default, the config has the learning rate as ``3e-4`` which is the same as the LoRA configs. For these experiments,
```

> we changed the learning rate from as high as 1e-3 to as low as 1e-5. To change the learning rate,

Reviewer: mega nit: should be able to use latexify this using something like :math:`1e^{-1}`

Author: Changed all to use math.

you can simply override the learning rate parameter using:

.. code-block:: bash

    tune run knowledge_distillation_single_device --config llama3_2/knowledge_distillation_single_device optimizer.lr=[LR]

Based on the results, the optimal learning rate changes depending on which metric you are optimizing for.

.. image:: /_static/img/kd-hyperparam-lr.png

Based on the loss graphs, all learning rates result in similar losses except for 1e-5, which has a higher KD and class loss.

Hyperparameter tuning: KD ratio
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In the config, we have the kd_ratio as 0.5, which gives even weightings to both the class and KD loss. In these experiments,
we look at the effects of different KD ratios, where 0 only uses the class loss and 1 only uses the KD loss.
Similar to changing the learning rate, the KD ratio can be adjusted using:

.. code-block:: bash

>     tune run knowledge_distillation_single_device --config llama3_2/knowledge_distillation_single_device kd_ratio=[KD_RATIO]

Reviewer: Same comment as above

Overall, the evaluation results are slightly better for higher KD ratios.

.. image:: /_static/img/kd-hyperparam-kd-ratio.png

Qwen2 1.5B to 0.5B
^^^^^^^^^^^^^^^^^^

> The KD recipe can also be applied to different model families as well. Here we look at the effect of KD when the number of

Reviewer: mega nit: 
```suggestion
The KD recipe can also be applied to different model families. Here we look at the effect of KD when the number of
```

> parameters between the teacher and student models are closer. For this experiment, we used Qwen2 1.5B and Qwen2 0.5B, which can be found in

Reviewer: nit:
```suggestion
parameters between the teacher and student models are closer. For this experiment, we used Qwen2 1.5B and Qwen2 0.5B, the configs for which can be found in
```

`qwen2/knowledge_distillation_single_device <https://github.com/pytorch/torchtune/blob/4234b78b914af23384ce0348f564e2119d107a96/recipes/configs/qwen2/knowledge_distillation_single_device.yaml>`_

> config. Here we see that alpaca_cleaned_dataset only improves truthful_qa performance and drops the metrics for the other evaluation tasks.

Reviewer: ```suggestion
config. Here we see that training on the alpaca cleaned dataset only improves truthful_qa performance and drops the metrics for the other evaluation tasks.
```

====================================================================

This guide will teach you about knowledge distillation (KD) and show you how you can use torchtune to distill a Llama3.1 8B model into Llama3.2 1B.
If you already know what knowledge distillation is and want to get straight to running your own distillation in torchtune,
you can jump to the `KD recipe in torchtune`_ tutorial.

.. grid:: 2

    .. grid-item-card:: :octicon:`mortar-board;1em;` What you will learn

      * What KD is and how it can help improve model performance
      * An overview of KD components in torchtune
      * How to distill from a teacher to student model using torchtune
      * How to experiment with different KD configurations

    .. grid-item-card:: :octicon:`list-unordered;1em;` Prerequisites

      * Be familiar with :ref:`torchtune<overview_label>`
      * Make sure to :ref:`install torchtune<install_label>`
      * Make sure you have downloaded the :ref:`Llama3 model weights<download_llama_label>`
      * Be familiar with :ref:`LoRA<lora_finetune_label>`

What is Knowledge Distillation?
-------------------------------

> `Knowledge Distillation <https://arxiv.org/pdf/1503.02531>`_ is is a widely used compression technique

Reviewer 2: ```suggestion
`Knowledge Distillation <https://arxiv.org/pdf/1503.02531>`_ is a widely used compression technique
```

that transfers knowledge from a larger (teacher) model to a smaller (student) model. Larger models have
more parameters and capacity for knowledge, however, this larger capacity is also more computationally
expensive to deploy. Knowledge distillation can be used to compress the knowledge of a larger model into
a smaller model. The idea is that performance of smaller models can be improved by learning from larger
model's outputs.

How does Knowledge Distillation work?
-------------------------------------

> Knowledge is transferred from the teacher to student model by training it on a transfer set and where the

Reviewer 2: ```suggestion
Knowledge is transferred from the teacher to student model by training it on a transfer set where the
```

  import torch.nn.functional as F

  class ForwardKLLoss(torch.nn.Module):
    def __init__(self, ignore_index: int = -100)
      super().__init__()
      self.ignore_index = ignore_index

    def forward(self, student_logits, teacher_logits, labels) -> torch.Tensor:
      # Implementation from https://github.com/jongwooko/distillm
      # Computes the softmax of the teacher logits
      teacher_prob = F.softmax(teacher_logits, dim=-1, dtype=torch.float32)
      # Computes the student log softmax probabilities
      student_logprob = F.log_softmax(student_logits, dim=-1, dtype=torch.float32)
      # Computes the forward KL divergence
      prod_probs = teacher_prob * student_logprob
      # Compute the sum
      x = torch.sum(prod_probs, dim=-1).view(-1)
      # We don't want to include the ignore labels in the average
      mask = (labels != self.ignore_index).int()
      # Loss is averaged over non-ignored targets
      return -torch.sum(x * mask.view(-1), dim=0) / torch.sum(mask.view(-1), dim=0)

There are some details omitted to simplify the computation, but if you'd like to know more,
you can see the implementation in :class:`torchtune.modules.loss.ForwardKLLoss`.

> By default, the KD configs use :class:`torchtune.modules.loss.ForwardKLWithChunkedOutputLoss` to reduce memory.

Reviewer 2: I think you need to add these APIs to the rst file [here](https://github.com/pytorch/torchtune/blob/4efd7fdb48677a4ac6d78f394be7b4ecfbabf7de/docs/source/api_ref_modules.rst?plain=1#L35). This way it'll render as an actual pointer to our API ref

Author: added. Thanks for the pointer.
